README.md
# Neighborhoods Yelp-Fusion Library

[![Build Status](https://travis-ci.org/neighborhoods/Yelp-Fusion.svg?branch=master)](https://travis-ci.org/neighborhoods/Yelp-Fusion)
[![Latest Stable Version](https://poser.pugx.org/neighborhoods/yelp-fusion/version)](https://packagist.org/packages/neighborhoods/yelp-fusion)
[![Total Downloads](https://poser.pugx.org/neighborhoods/yelp-fusion/downloads)](https://packagist.org/packages/neighborhoods/yelp-fusion)
[![License](https://poser.pugx.org/neighborhoods/yelp-fusion/license)](https://packagist.org/packages/neighborhoods/yelp-fusion)

A library that interfaces with the Yelp Fusion (v3) API.

## Installation

#### Requirements
* PHP >= 5.6
* Yelp Fusion API credentials

#### Installation

`composer require neighborhoods/yelp-fusion`

## Usage

#### Getting API key

Yelp Fusion API uses private API Keys to authenticate requests. In order to set up your access to Yelp Fusion API, you need to create an app with Yelp.

See [Yelp's authentication documentation](https://www.yelp.com/developers/documentation/v3/authentication) for instructions.

#### Search

```php
$params = [
    'categories' => 'arts',
    'latitude'   => 41.879562,
    'longitude'  => -87.624205,
    'radius'     => 16093,
];

$yelpData = $this->yelp->search($params, $apiKey);
$businesses = $yelpData->businesses;
```

See [Yelp's business search documentation](https://www.yelp.com/developers/documentation/v3/business_search) for the list of available search parameters and the shape of the results object.


CHANGELOG.md
# Release Notes

## v2.0.0 (2018-07-06)

### Removed
- Removed `getBearerTokenObject` method. The token endpoint has been deprecated.
Yelp now uses persistent API keys.


tests/YelpTest.php
<?php
/**
 * Test suite for the Yelp library
 *
 * PHP version 5.6
 *
 * @category Library
 * @package  Neighborhoods\YelpFusion
 * @author   Matt Stypa <matt.stypa@55places.com>
 * @author   Rashaud Teague <rashaud.teague@55places.com>
 * @license  MIT https://opensource.org/licenses/MIT
 * @link     https://github.com/neighborhoods/Yelp-Fusion
 */

namespace Neighborhoods\YelpFusion\test;

use Exception;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Middleware;
use GuzzleHttp\Exception\RequestException;
use PHPUnit\Framework\TestCase;
use Neighborhoods\YelpFusion\Yelp;

class YelpTest extends TestCase
{
    protected $yelp;
    protected $requestHeaders = [
        'cache-control' => 'no-cache',
    ];

    public function testSearchResponse()
    {
        $yelp = new Yelp(
            $this->getMockHandler(
                new Response(200, [], '{"test": "success"}')
            )
        );
        $response = $yelp->search([], 'BEARER TOKEN');

        $this->assertEquals('object', gettype($response));
        $this->assertEquals('success', $response->test);
    }

    public function testSearchRequest()
    {
        $params = [
            'categories' => 'arts',
            'latitude' => '41.891564',
            'longitude'=> '-87.611906',
        ];
        $requests = [];
        $yelp = new Yelp(
            $this->getMockHandler(
                new Response(200, [], '{"test": "success"}'),
                $requests
            )
        );
        $yelp->search($params, 'BEARER TOKEN');

        $requestHeaders = $requests[0]['request']->getHeaders();
        $requestQuery = $requests[0]['request']->getUri()->getQuery();

        parse_str($requestQuery, $parsedRequestQuery);

        $this->assertEquals(Yelp::HTTP_GET, $requests[0]['request']->getMethod());
        $this->assertEquals($params, $parsedRequestQuery);

        foreach ($this->requestHeaders as $headerKey => $headerValue) {
            $this->assertEquals($headerValue, $requestHeaders[$headerKey][0]);
        }
    }

    public function testSearchThrowsOnCommunicationError()
    {
        $yelp = new Yelp(
            $this->getMockHandler(
                new RequestException('error', new Request('GET', 'test'))
            )
        );

        $this->expectException(Exception::class);

        $yelp->search([], 'BEARER TOKEN');
    }

    public function testSearchThrowsOnNon200Response()
    {
        $yelp = new Yelp(
            $this->getMockHandler(
                new Response(201, [], '{"test": "success"}')
            )
        );

        $this->expectException(Exception::class);

        $yelp->search([], 'BEARER TOKEN');
    }

    public function testSearchThrowsOnNonJSONResponse()
    {
        $yelp = new Yelp(
            $this->getMockHandler(
                new Response(200, [], 'This is not JSON')
            )
        );

        $this->expectException(Exception::class);

        $yelp->search([], 'BEARER TOKEN');
    }

    protected function getMockHandler($response, &$requests = [])
    {
        $history = Middleware::history($requests);
        $handler = new MockHandler([$response]);
        $stack = HandlerStack::create($handler);
        $stack->push($history);

        return $stack;
    }
}


src/Yelp.php
<?php
/**
 * A library that interfaces with the Yelp Fusion (v3) API
 *
 * PHP version 5.6
 *
 * @category Library
 * @package  Neighborhoods\YelpFusion
 * @author   Matt Stypa <matt.stypa@55places.com>
 * @author   Rashaud Teague <rashaud.teague@55places.com>
 * @license  MIT https://opensource.org/licenses/MIT
 * @link     https://github.com/neighborhoods/Yelp-Fusion
 */

namespace Neighborhoods\YelpFusion;

use Exception;
use GuzzleHttp\Client;

/**
 * Class Yelp
 *
 * @package Neighborhoods\YelpFusion
 */
class Yelp
{
    const BASE_URI = 'https://api.yelp.com/';
    const ENDPOINT_SEARCH = 'v3/businesses/search';
    const TIMEOUT_IN_SECONDS = 5;
    const HTTP_GET = 'GET';
    const HTTP_POST = 'POST';

    protected $requestHeaders = [
        'cache-control' => 'no-cache',
    ];

    protected $guzzle;

    /**
     * Yelp constructor.
     *
     * @param null $handler, optional Guzzle handler to assist for testing
     */
    public function __construct($handler = null)
    {
        $options = [
            'base_uri' => self::BASE_URI,
            'timeout'  => self::TIMEOUT_IN_SECONDS,
            'headers'  => $this->requestHeaders,
        ];

        if ($handler) {
            $options['handler'] = $handler;
        }

        $this->guzzle = new Client($options);
    }

    /**
     * Gets the search results
     *
     * @param $terms
     * @param $apiKey
     * @return object
     * @throws Exception
     */
    public function search($terms, $apiKey)
    {
        return $this->parseResponse(
            $this->guzzle->get(
                self::ENDPOINT_SEARCH,
                [
                    'headers' => [
                        'authorization' => 'Bearer ' . $apiKey,
                    ],
                    'query'   => $terms,
                ]
            )
        );
    }

    /**
     * Parses the response (decodes the json response)
     *
     * @param $response
     * @return object
     * @throws Exception
     */
    protected function parseResponse($response)
    {
        if ($response->getStatusCode() !== 200) {
            throw new Exception('Invalid response');
        }

        $parsedResponse = json_decode($response->getBody());

        if (!$parsedResponse) {
            throw new Exception('Unable to parse response');
        }

        return $parsedResponse;
    }
}
